Robustify Rayleigh Bernard two phase test

lindsayad opened the pull request:
Refs failures from the combination of #26462 and #26227

To avoid negative viscosity interpolations I shrank the growth factor from 2 to 1.5. Once I did this I got a CSVDiff which signaled to me that the original result was solved to too loose a tolerance. So I removed the steady state detection, which has potential issues with it (see #26312), and extended the `end_time` until we hit a fairly tight absolute tolerance

lindsayad committed:
Robustify Rayleigh Bernard two phase test

Refs failures from the combination of #26462 and #26227

To avoid negative viscosity interpolations I shrank the growth
factor from 2 to 1.5. Once I did this I got a CSVDiff which signaled
to me that the original result was solved to too loose a tolerance.
So I removed the steady state detection, which has potential issues
with it (see #26312), and extended the `end_time` until we hit
a fairly tight absolute tolerance

lindsayad commented:
FYI @loganharbour @tanoret

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2030029/) on 2b34c70 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/26658/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2030044/) on 2b34c70 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>41d8b3</th>
      <th colspan=4>#26658 2b34c7</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.23%</td>
      <td>85.22%</td>
      <td>-0.01%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7190</td>
      <td>7189</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1246</td>
      <td>1247</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26658/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26658/coverage/fluid_properties)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>41d8b3</th>
      <th colspan=4>#26658 2b34c7</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.73%</td>
      <td>84.75%</td>
      <td>+0.02%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>14564</td>
      <td>14568</td>
      <td>+4</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2625</td>
      <td>2621</td>
      <td>-4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26658/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26658/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/26658/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad commented:
the diff is quite large; it makes me concerned there is something genuinely wrong either with distributed mesh or with threads

tanoret commented:
> the diff is quite large; it makes me concerned there is something genuinely wrong either with distributed mesh or with threads

This is mostly because the problem admits multiple solutions, which makes it a bit sensitive to the initial seed in the solution. We can always invert gravity and make the flow stagnate to avoid multiple solutions, but then the problem is not very interesting anymore. What do you think?

lindsayad commented:
I think that if the test is driving to a steady-state, the test should only have a single unique solution. We want the test to be deterministic; we don't want the result to be influenced by small perturbations in the order of floating point arithmetic. Or if the test input admits multiple solutions, then we need to modify what we are testing

tanoret commented:
Actually, in steady-state the test admits multiple solutions; rising plume by the center of the domain and down by the sides, rising plume by the sides of the domain and down by the center, and these permutations exists for all the eddies formed (e.g., https://journals.aps.org/pre/abstract/10.1103/PhysRevE.105.055106). Most natural-convection-driven problems for two-phase flow will admit multiple solutions. That is why I think we need a test with a high tolerance to encompass to guarantee no regression  in the multiple solutions that may exist in the problem. If you guys don't like this, we can always invert gravity, stratify the flow, and go into a unique solution case. I am good with any of these cases

lindsayad commented:
> That is why I think we need a test with a high tolerance

I don't really understand what this means. This sounds like a problem that has just been solved incompletely. A problem solved incompletely seems like it could be a solve that is on its way to solution 1, or to solution 2, or since it's solved incompletely a superposition of them. How can you be certain that the solver will repeatably go to the same incompletely solved solution? As soon as we updated PETSc we were perturbed to a different state, which is why next is red https://civet.inl.gov/branch/6695/. So the loose tolerance from before was not leading to robust, repeatable final test results

lindsayad committed:
Working input with 1, 4, and 8 threads

lindsayad committed:
Create WCNSFV2PSlipVelocityFunctorMaterial

lindsayad committed:
Avoid zero norm of velocity in NSFVDispersePhaseDragFunctorMaterial

lindsayad committed:
Test too heavy to run with dbg/valgrind

lindsayad commented:
the results in this test are very cool by the way. very cool new functionality. now we just have to figure out what to do with civet

tanoret commented:
> > a test with a high tolerance
> 
> I think I may have interpreted this wrong. I assumed you meant a loose solve tolerance, but perhaps you meant a loose CSVDiff tolerance? That does make more sense to me

Yes. That is the one, loose CSV tolerance. Should we make it looser? Is that possible? If we break something up, it should still show up if the tolerance is small enough

lindsayad committed:
Go with odd number of elements

I noticed that when a diff occurred, looking at vel_y, it was a
mirror plane switch across the x = L/2 line. So by going with an
odd number of elements this doesn't occur. I tested this with
parallel and/or thread sweeps with and without distributed mesh
and could not get it to fail

lindsayad commented:
see my newest commit which just changes nx and ny to 11 instead of 10. This seems to lead to these repeatable solutions for vel-x and vel-y
![Screenshot 2024-01-29 at 1 40 42 PM](https://github.com/idaholab/moose/assets/10248304/df186ad9-2bc3-411b-8c87-5994db77cec5)
![Screenshot 2024-01-29 at 1 41 16 PM](https://github.com/idaholab/moose/assets/10248304/3911daa8-8968-48af-bfcd-129348990905)

tanoret commented:
Yes. Looks great! Nice idea introducing anisotropy in the mesh

moosebuild commented:
Job [Framework increased sparsity](https://civet.inl.gov/job/2032983/) on 91c6103 : invalidated by @lindsayad

grmnptr commented on modules/navier_stokes/src/functormaterials/WCNSFV2PSlipVelocityFunctorMaterial.C:
```diff
@@ -0,0 +1,141 @@
+  params.addParam<PostprocessorName>(
+      "force_postprocessor", 0, "A postprocessor whose value is multiplied by the body force");
+  params.addParam<RealVectorValue>(
+      "force_direction", RealVectorValue(1, 0, 0), "Gravity acceleration vector");
```
```suggestion
      "force_direction", RealVectorValue(1, 0, 0), "Gravitational acceleration vector");
```

lindsayad replied:
I'll apply this in #26683

grmnptr replied:
Sounds good!

grmnptr reviewed (approved):
Looks good to me!

lindsayad commented:
failures are unrelated
